src/passes/DeadArgumentElimination.cpp

```diff
@@ -176,49 +189,83 @@ struct DAE : public Pass {
+      if (!iteration(module, infoMap)) {
         break;
       }
     }
   }
 
-  bool iteration(Module* module) {
+  bool iteration(Module* module, DAEFunctionInfoMap& infoMap) {
     allDroppedCalls.clear();
 
-    DAEFunctionInfoMap infoMap;
-    // Ensure they all exist so the parallel threads don't modify the data
-    // structure.
-    for (auto& func : module->functions) {
-      infoMap[func->name];
+#if DAE_DEBUG
+    // Enable this path to mark all contents as stale at the start of each
+    // iteration, which can be used to check for staleness bugs. Note, though,
+    // that staleness bugs can easily cause serious issues with validation (e.g.
+    // if data is stale we may miss that there is an additional caller, that
+    // prevents refining argument types etc.), so this may not be terribly
+    // helpful.
+    if (getenv("ALWAYS_MARK_STALE")) {
+      for (auto& [_, info] : infoMap) {
+        info.markStale();
+      }
     }
+#endif
+
     DAEScanner scanner(&infoMap);
     scanner.walkModuleCode(module);
-    for (auto& curr : module->exports) {
-      if (curr->kind == ExternalKind::Function) {
-        infoMap[curr->value].hasUnseenCalls = true;
-      }
-    }
     // Scan all the functions.
     scanner.run(getPassRunner(), module);
     // Combine all the info.
+    struct CallContext {
+      Call* call;
+      Function* func;
+    };
     std::map<Name, std::vector<Call*>> allCalls;
     std::unordered_set<Name> tailCallees;
-    for (auto& [_, info] : infoMap) {
+    std::unordered_set<Name> hasUnseenCalls;
+    // Track the function in which relevant expressions exist. When we modify
+    // those expressions we will need to mark the function's info as stale.
+    std::unordered_map<Expression*, Name> expressionFuncs;
+    for (auto& [func, info] : infoMap) {
       for (auto& [name, calls] : info.calls) {
         auto& allCallsToName = allCalls[name];
         allCallsToName.insert(allCallsToName.end(), calls.begin(), calls.end());
+        for (auto* call : calls) {
+          expressionFuncs[call] = func;
+        }
       }
       for (auto& callee : info.tailCallees) {
         tailCallees.insert(callee);
       }
-      for (auto& [name, calls] : info.droppedCalls) {
-        allDroppedCalls[name] = calls;
+      for (auto& [call, dropp] : info.droppedCalls) {
+        allDroppedCalls[call] = dropp;
+      }
+      for (auto& name : info.hasUnseenCalls) {
+        hasUnseenCalls.insert(name);
+      }
+    }
+    // Exports are considered unseen calls.
+    for (auto& curr : module->exports) {
+      if (curr->kind == ExternalKind::Function) {
+        hasUnseenCalls.insert(curr->value);
       }
     }
 
-    // Track which functions we changed, and optimize them later if necessary.
+    // Track which functions we changed that are worth re-optimizing at the end.
+    // (This is not the set of all changed functions, as some changes are
+    // trivial and not likely to lead to optimization opportunities.)
```

Comment:
It would be good to give the variable a more descriptive name if it is no longer a set of all changed functions.

Reply:
Done. Earlier I was a little worried it would end up causing changes that would obscure the core modifications, but I agree it's clearer now.